server/server.go

```diff
@@ -101,35 +103,48 @@ func (srv *Server) init() {
 		if srv.handler == nil {
 			srv.handler = http.DefaultServeMux
 		}
+		// Setup health checks, /healthz route is taken by health checks by default.
+		// Note: App Engine Flex uses /_ah/health by default, which can be changed
+		// in app.yaml. We may want to do an auto-detection for flex in future.
+		const healthPrefix = "/healthz/"
+
+		mux := http.NewServeMux()
+		mux.HandleFunc(healthPrefix+"liveness", health.HandleLive)
+		mux.Handle(healthPrefix+"readiness", &srv.healthHandler)
+		h := srv.handler
+		if srv.reqlog != nil {
+			h = requestlog.NewHandler(srv.reqlog, h)
+		}
+		h = &ochttp.Handler{
+			Handler:          h,
+			IsPublicEndpoint: true,
+		}
+		mux.Handle("/", h)
+		srv.wrappedHandler = mux
 	})
 }
 
 // ListenAndServe is a wrapper to use wherever http.ListenAndServe is used.
-// It wraps the passed-in http.Handler with a handler that handles tracing and
+// It wraps the http.Handler provided to New with a handler that handles tracing and
 // request logging. If the handler is nil, then http.DefaultServeMux will be used.
 // A configured Requestlogger will log all requests except HealthChecks.
 func (srv *Server) ListenAndServe(addr string) error {
 	srv.init()
+	return srv.driver.ListenAndServe(addr, srv.wrappedHandler)
+}
 
-	// Setup health checks, /healthz route is taken by health checks by default.
-	// Note: App Engine Flex uses /_ah/health by default, which can be changed
-	// in app.yaml. We may want to do an auto-detection for flex in future.
-	const healthPrefix = "/healthz/"
-
-	mux := http.NewServeMux()
-	mux.HandleFunc(healthPrefix+"liveness", health.HandleLive)
-	mux.Handle(healthPrefix+"readiness", &srv.healthHandler)
-	h := srv.handler
-	if srv.reqlog != nil {
-		h = requestlog.NewHandler(srv.reqlog, h)
-	}
-	h = &ochttp.Handler{
-		Handler:          h,
-		IsPublicEndpoint: true,
+// ListenAndServeTLS is a wrapper to use wherever http.ListenAndServeTLS is used.
+// It wraps the http.Handler provided to New with a handler that handles tracing and
+// request logging. If the handler is nil, then http.DefaultServeMux will be used.
+// A configured Requestlogger will log all requests except HealthChecks.
+func (srv *Server) ListenAndServeTLS(addr, certFile, keyFile string) error {
+	// Check if the driver implements the optional interface.
+	tlsDriver, ok := srv.driver.(driver.TLSServer)
```

Comment:
Is this check necessary? Would the compiler give a similar error message?

Reply:
No, because one can instantiate a `Server` with a `driver.Server`; `driver.TLSServer` is an additional interface that drivers can choose whether to implement. For example, the test has two drivers, one that supports TLS and one that doesn't.